src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -1400,482 +1140,148 @@ class ProgResolutionDirectionalFsc : public XmippProgram
-
-    	double cutoff;
-    	cutoff = sampling/resol;
-    	Matrix2D<double> indexesFourier, indexesFourier2;
-    	MultidimArray<std::complex<double>> f1, f2;
-    	fscShell(FT1, FT2, freq_fourier_x, freq_fourier_y, freq_fourier_z, freqMap, m1sizeX, indexesFourier, indexesFourier2, cutoff, f1, f2);
-
 
-    	bool angledependence = false;
-    	double dresfsc, lastcross = 1e38;
-    	FileName fnmd;
-    	if (doCrossValidation || angledependence)
-		{
-    		size_t count = 0;
-    		generateDirections(angles, true);
-			double angCon, cross;
-			size_t objId;
-			MetaData mdcrossval;
-
-			for (double myangle = 10; myangle < 41; myangle = myangle + 1)
-			{
-				angCon = myangle*PI/180;
-				cross = 0;
-				aniParam.initZeros(m1sizeX/2+1);
-				for (size_t k = 0; k<angles.mdimx; k++)
-				{
-					double rot = MAT_ELEM(angles, 0, k);
-					double tilt = MAT_ELEM(angles, 1, k);
+		//This read the data and applies a fftshift to in the next step compute the fft
+    	prepareData(half1, half2);
 
-					if (doCrossValidation)
-					{
-						weights(indexesFourier, indexesFourier2, rot, tilt, myangle,
-									 f1, f2, cross);
-					}
-					if (angledependence)
-					{
-						fscDir(FT1, FT2, sampling, freq_fourier_x, freq_fourier_y, freq_fourier_z, freqMap, freq, fsc, 0.5,
-											m1sizeX, m1sizeY, m1sizeZ, rot, tilt, myangle,  dresfsc, thrs);
+		//Computing the FFT
+		FourierTransformer transformer2(FFTW_BACKWARD), transformer1(FFTW_BACKWARD);
+		transformer1.setThreadsNumber(Nthreads);
+		transformer2.setThreadsNumber(Nthreads);
 
-						anistropyParameterSimple(fsc, aniParam, thrs);
-					}
+        transformer1.FourierTransform(half1, FT1, false);
+     	transformer2.FourierTransform(half2, FT2, false);
 
-				}
-				if (angledependence)
-				{
-					//ANISOTROPY CURVE
-					aniParam /= (double) angles.mdimx;
-					MetaData mdani;
-					fnmd = fn_fscmd_folder+formatString("AniIterAngle_%i.xmd", count);
-					saveAnisotropyToMetadata(mdani, freq, aniParam, fnmd);
-					++count;
-				}
+		// Defining frequencies freq_fourier_x,y,z and freqMap
+		// The number of frequencies in each shell freqElem is determined
+        defineFrequencies(FT1, phalf1);
 
-				objId = mdcrossval.addObject();
-				mdcrossval.setValue(MDL_ANGLE_Y, myangle, objId);
-				mdcrossval.setValue(MDL_SUM, sqrt(cross), objId);
-				if (cross<lastcross)
-				{
-					lastcross = cross;
-					ang_con = myangle;
+		half1.clear();
+		half2.clear();
 
-				}
-			}
-			mdcrossval.write(fn_fscmd_folder+"crossValidation.xmd");
-		}
+		// Storing the shell of both maps as vectors global
+		// The global FSC is also computed
+		MultidimArray<double> freq;
+		double resol, resInterp;
+		arrangeFSC_and_fscGlobal(sampling, thrs, resInterp, freq);
 
-    	if (doCrossValidation == true)
-    		std::cout << "The best cone angle is " << ang_con << std::endl;
-    	else
-    		std::cout << "The chosen cone angle is " << ang_con << std::endl;
+		std::cout << " " << std::endl;
+		FT2.clear();
```

Comment:
I don't see FT2 used anywhere before you clear it.

Reply:
it is used in a function, but I prefer to keep it in the main function, run()